oak_runtime/src/node/http/client.rs

```diff
@@ -0,0 +1,280 @@
+    pub fn new(
+        node_name: &str,
+        config: HttpClientConfiguration,
+        root_ca_path: String,
+    ) -> Option<Self> {
+        let http_client = create_client(root_ca_path).ok()?;
+        let node_privilege = get_privilege(config.uri.clone());
+        Some(HttpClientNode {
+            node_name: node_name.to_string(),
+            uri: config.uri,
+            http_client,
+            node_privilege,
+        })
+    }
+
+    /// Main loop that handles HTTP invocations from the `handle`, sends HTTP requests to an
+    /// external HTTP service and writes HTTP responses back to the invocation channel.
+    async fn handle_loop(&mut self, runtime: RuntimeProxy, handle: Handle) -> Result<(), OakError> {
+        // Create a [`Receiver`] to receive the HTTP request.
+        let receiver = Receiver::<Invocation>::new(ReadHandle { handle });
+        loop {
+            debug!("Waiting for HTTP invocation");
+            // Read an HTTP invocation from the [`Receiver`].
+            let invocation = receiver.receive(&runtime).map_err(|error| {
+                match error {
+                    OakError::OakStatus(OakStatus::ErrTerminated) => {
+                        debug!("HTTP client node is terminating.")
+                    }
+                    OakError::OakStatus(OakStatus::ErrChannelClosed) => {
+                        info!("HTTP invocation channel closed")
+                    }
+                    _ => error!("Couldn't receive the invocation: {:?}", error),
+                }
+                error
+            })?;
+
+            let result = self.process_invocation(&runtime, &invocation).await;
+            info!("Invocation processing finished: {:?}", result);
+            if result.is_err() {
+                warn!("Error encountered: {:?}", result);
+            }
+            invocation.close(&runtime);
+        }
+    }
+
+    /// Process an HTTP invocation for an external HTTP service.
+    async fn process_invocation(
+        &mut self,
+        runtime: &RuntimeProxy,
+        invocation: &Invocation,
+    ) -> Result<(), OakError> {
+        // TODO(#1717): Update HTTP client metrics
+
+        // Define a closure to ease logging the errors
+        let send_error = |code, msg| {
+            invocation.send_error(code, msg, runtime);
+        };
+
+        // Receive a request from the invocation channel.
+        let request = invocation.receive_request(&runtime).map_err(|error| {
+            send_error(
+                http::StatusCode::INTERNAL_SERVER_ERROR,
+                "Failed to read request",
+            );
+            error!(
+                "Couldn't read HTTP request from the invocation: {:?}",
+                error
+            );
+            error
+        })?;
+        debug!("Incoming HTTP request: {:?}", request);
+
+        let mut hyper_request_builder = hyper::Request::builder()
+            .method(&request.method[..])
+            .uri(request.uri);
```

Comment:
I am not sure I understand why there is a request URI here, but there is also one that gets specified at node creation time, and how they are related to each other, and to the node privilege. Could you point me to the relevant code?

Reply:
This is something that I needed to discuss :) 
I am not sure what is the best way to send the requests to the HTTP client. I think to follow the same pattern as in the gRPC client, we could have the base URI in the HTTP client, and each request would only contain a relative URI. 
The other alternative is to let the incoming HTTP requests to have the absolute URI, but check those URIs against the base URI in the HTTP client. I think generally we want this kind of checks to be done more strictly, or via the IFC. So, perhaps the first approach is the better one. The only problem I see with that is that it might be confusing that sometimes we use the `uri` in the `HttpRequest` to be the absolute URI, but in this case it has to be the relative URI. WDYT?

If we can go for the second alternative, we could perhaps start the HTTP client with an allow-list of authorities (instead of having a dedicated client per authority as we have for gRPC). Incoming requests come with absolute URIs, but are only handled if the authority matches one of the authorities in the allow-list. With this approach, we'd probably need disjunctions too.

Reply:
For consistency with the gRPC client, I would favour only accepting relative URIs. The gRPC method is effectively a relative URI.

Reply:
Accepting relative URIs in the request would probably not be very ergonomic for callers (also error prone). I suggest allowing the caller to specify the full URI, but then the node would check whether the schema is HTTPS, and whether the host matches the one defined in the node configuration.

Thinking more long-term, I actually think that we should get rid of the hardcoded host in the configuration eventually. An HTTP client pseudo-node may be created with an arbitrary label (and no config), and at creation time we would chech whether the label is either public or a TLS cert one (for now let's ignore connectives).

If the HTTP client node is created with a public label, then it will allow arbitrary requests to any URL for any host. Of course only nodes that are public or can declassify to public will be able to use it, but I think that's probably an easy win that covers a lot of nodes that just fetch some external resource at startup.

If the HTTP client node is created with a TLS cert label, then at runtime it will check whether the host of the incoming request matches the label (and privilege), and only proceed if the node is able to effectively declassify the request before sending it out.

Reply:
If the HTTP client node automatically gets the privilege to declassify based on its assigned label, it is equivalent to just adding a TLS endpoint Top tag and always giving that privilege to the HTTP client node. I think the current approach of restricting the authority is safer. If a single node needs to connect to multiple different servers, we could perhaps change the HTTP client node configuration to have a list of authorities (rather than just a single one).

Reply:
That is an interesting observation :) I prefer a node with  theTop privilege and an allowlist of authorities. It gives more flexibility.